File: src/main/scala/com/zooplus/jacekb/learningTime/akka/pi/Worker.scala
package com.zooplus.jacekb.learningTime.akka.pi

import akka.actor.{Actor, ActorLogging}
import com.zooplus.jacekb.learningTime.akka.pi.Commons.{Result, Work}
import akka.cluster.Cluster

/**
 * Created with IntelliJ IDEA.
 * User: jacek_bilski
 * Date: 15/05/13
 * Time: 08:21
 */

class Worker extends Actor with ActorLogging {

	println("Creating worker")

	val role = "worker"

	val piCalculator = new PiCalculator
	val cluster = Cluster(context.system)

	override def preStart() {
		log.info("Worker starting")
		cluster.subscribe(self, classOf[Work])
	}

	override def postStop() {
		cluster.unsubscribe(self)
	}

	def receive = {
		case Work(start, nrOfElements) ⇒
			//log.info("Got Work(" + start + ", " + nrOfElements + ")")
			sender ! Result(piCalculator.calculatePiFor(start, nrOfElements)) // perform the work
	}
}

File: src/main/java/com/zooplus/jacekb/learningTime/akka/pi/java/Worker.java
package com.zooplus.jacekb.learningTime.akka.pi.java;

import akka.actor.AbstractActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;

import java.math.BigDecimal;

/**
 * Created with IntelliJ IDEA.
 * User: jacek_bilski
 * Date: 18.09.14
 * Time: 09:58
 */
public class Worker extends AbstractActor {

	private LoggingAdapter log = Logging.getLogger(getContext().system(), this);

	private PiCalculator piCalculator;

	public Worker() {
		receive(ReceiveBuilder.match(Work.class, this::calculate).build());
		piCalculator = new PiCalculator();
	}

	private void calculate(Work work) {
		log.info("Got Work(" + work.getStart() + ", " + work.getNrOfElements() + ")");
		BigDecimal value = piCalculator.calculatePiFor(work.getStart(), work.getNrOfElements());
		sender().tell(new Result(value), self());
	}
}
